### PR TITLE
SAK-29405: Entity Providers for Course Management Entities

### PR DESCRIPTION
**Pull request by cschauer**
Cmprovider provides a RESTful API for updating course management data.
It can be used as an alternative to csv upload or manually adding cm
data to the database. It's built using the entity broker system and can
be accessed at /direct.

### CODE REVIEW

**Comment by cschauer** [comment]
This is a pretty big pull request. I'd love to get some feedback on this. When I talked to Neil about getting this project in 11, he mentioned that attending a core team call would be a good way to get feedback on my code. I'd be more than happy to do that if others think it's a good idea. Just let me know when and how.

Also, I have a few questions about the best way to handle some things:

1) Permissions. Right now, you have to be an admin to access the API. I think this makes the most sense, but I'd appreciate some guidance if there's a better way.

2) Updating sites when enrollments or memberships change. We use enrollments here at Texas State and don't actually touch memberships at all. In order to update sites when an enrollment changes, I'm just calling save on the realm since I know the save function goes through a process of updating site participants from the CM tables. Obviously, this isn't very efficient. If someone more familiar with how the course management stuff ties into realms wouldn't mind talking with me about this I'd greatly appreciate it.

3) Testing. Unit tests didn't really make sense for this, since it's basically a wrapper for the Course Management API. I have some functional tests written in Go that send requests to this API, but those can't really be included here. Is it okay that there aren't any tests included here? I could post my functional tests in a separate repo. Any suggestions?

**Comment by steveswinsburg** [comment]
This is a pretty massive change that can be made a lot smaller by getting rid of the redundant entitybroker interfaces that you are creating. Just collapse that into the provider classes.

Also, I'm not keen on those marker annotations going in. You could maybe use the javax.validation annotations to do the same thing without the custom code, but its a pretty non standard pattern in Sakai.

Also the DateUtils class has a fixed date format. Is that a fixed field or is it locale dependant?

**Comment by cschauer** [comment]
Thanks for the feedback Steve. I will remove the interfaces and update the Date parsing. Yes, it is a fixed field. When I wrote these originally for our use locally, it was for 2.9 with Java 6. Java 8 looks to have some nice date formatting libraries, but the stuff for Java 6 was confusing and hard to work with. I just went with the fixed value for simplicity at the time, but I will revisit it.

With regard to the annotations for validation, what don't you like about it? I'd prefer to not rewrite all the validation logic.

**Comment by cschauer** [comment]
The java 8 DateTimeFormatter is nice, but using it would mean doing some messy conversion between the new java.time.LocalDate and java.util.Date since the Course Management API uses java.util.Date. I'm not sure I want to mess with that for such a simple case like this.

I'm inclined to leave it as is. YYYY-MM-DD follows the ISO-8601 standard. I could add support for YYYYMMDD as well since dashes are optional for ISO-8601. Thoughts?

**Comment by steveswinsburg** [comment]
I'm ok with the dates, just wanted a second look.

Re the annotations, it's just that it is a non standard pattern in Sakai
when there are plenty of more common alternatives and they are specific to
this bit of code. If the annotations were available in kernel then this
might be ok, but not for just this bit.

sent from my mobile device

**Comment by cschauer** [comment]
Fair enough. What is the standard pattern for entity validation in Sakai? Most of the entity providers I looked at are doing their validation inline and repeating it for each request. That was something I wanted to avoid.

Would it be ok if I replaced my custom annotations with ones from javax.validation? To make it reusable, I could add a Validateable capability to entitybroker. If a provider implements this interface then entities passed to create/update entity methods will be validated (using javax.validation) before those methods are called.

**Comment by steveswinsburg** [comment]
The current pattern has just been to validate the fields where used. More
verbose, but at least other people who need to maintain it understand
what's happening without having to peel back many layers of abstraction.

My approach to entity providers is just to have inner classes for the DTOs
that are returned. The classes like SectionData are not used anywhere else,
no use exposing it as an API. If it's just a wrapper around the base API
class (Section), I would just include it as an inner class within the
entity provider that it is used.

This whole PR can be reduced to the 8 entity provider classes as well as
the pom and component xmls which would make this super easy to maintain:

AcademicSessionEntityProvider, CanonicalCourseEntityProvider,
CourseOfferingEntityProvider, CourseSetEntityProvider,
EnrollmentEntityProvider, EnrollmentSetEntityProvider,
SectionCategoryEntityProvider, SectionEntityProvider

CmProviderHibernateService could maybe go into the main API but it's fine
to be where it is.

I value the quest for elegant code, but in this case, compact is better.

**Comment by cschauer** [comment]
You're right, I went a little overboard with spreading it out so much. I removed the api and pack modules and condensed everything into a single package.  I don't agree with making the data classes inner classes because a lot of those classes might be shared among the providers. For example, MembershipData is shared among a few providers. Most of the course management objects are linked together in some way. In the future, I'd like to add things like actions to get all enrollments on an enrollment set or to find course offerings in a course set. There's still a lot of functionality in the Course Management API that could be exposed.

I feel strongly about keeping the validation tied to the data objects. I've given this a lot of thought after reading your feedback. I think tying validation to each field of the data objects has merit too.  For example, I'd really like to add the ability to send PATCH requests through the entitybroker. That way if a title changes on a CourseOffering you could just send a PATCH to /direct/cm-course-offering/eid.json with { title: "New Title" } for the body. If there's a way to validate each field individually, then it's easy to add validation for this new action. If there isn't, then that means repeating more validation code and the code starts to get fragile.

I understand your point about readability. The validation is somewhat hidden in the base class and it isn't entirely clear from looking at the providers themselves that the data is getting validated or where. I think I can improve this without rewriting the whole thing.

**Comment by steveswinsburg** [comment]
If the classes are going to be used outside of those entityproviders then by all means expose them. The PATCH approach sounds good, not sure if a change to EB would be required to support that though.

But yeah if you can either get the validation annotations into kernel, they can definitely stay, or maybe use common ones, this would be ok.

Question: If the validation fails, does a user of that annotation get control over the response message?

File: cmprovider/pom.xml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-lang3</artifactId>
+      <version>3.3.2</version>
+    </dependency>
+    <dependency>
+      <groupId>commons-logging</groupId>
+      <artifactId>commons-logging</artifactId>
+      <version>1.1.3</version>
```
Comment (buckett) [inline cmprovider/pom.xml]: You shouldn't specify the version here as it will come from the parent POM.

File: cmprovider/pom.xml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      <artifactId>sakai-kernel-api</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.sakaiproject.kernel</groupId>
+      <artifactId>sakai-component-manager</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-core</artifactId>
+      <version>${sakai.spring.version}</version>
```
Comment (buckett) [inline cmprovider/pom.xml]: Should drop version.

File: cmprovider/pom.xml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-jdbc</artifactId>
+      <version>${sakai.spring.version}</version>
```
Comment (buckett) [inline cmprovider/pom.xml]: Should drop version.

File: cmprovider/pom.xml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.springframework</groupId>
+      <artifactId>spring-orm</artifactId>
+      <version>${sakai.spring.version}</version>
```
Comment (buckett) [inline cmprovider/pom.xml]: Should drop version.

File: cmprovider/pom.xml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.sakaiproject.edu-services.course-management</groupId>
+      <artifactId>coursemanagement-hibernate-impl</artifactId>
+      <version>1.2.3</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.commons</groupId>
+      <artifactId>commons-lang3</artifactId>
+      <version>3.3.2</version>
```
Comment (buckett) [inline cmprovider/pom.xml]: commons-lang3 is provided and so shouldn't have an explicit version.

File: cmprovider/pom.xml [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    </dependency>
+    <dependency>
+      <groupId>commons-logging</groupId>
+      <artifactId>commons-logging</artifactId>
+      <version>1.1.3</version>
+    </dependency>
+    <dependency>
+      <groupId>org.hibernate</groupId>
+      <artifactId>hibernate-core</artifactId>
+      <version>${sakai.hibernate.version}</version>
```
Comment (buckett) [inline cmprovider/pom.xml]: Drop the version.

**Comment by steveswinsburg** [comment]
Hey @cschauer is this one still good to go?

**Comment by cschauer** [comment]
Hi Steve. Yes, this is good to go. We're actually releasing this exact version into production here next Saturday. We've been using a previous version to sync course information with Banner since August.

**Comment by cschauer** [comment]
@steveswinsburg Thanks for the review and assistance with this PR.

**Pull request merged**